usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateAgencyLangAndFeedInfoFeedLangMatch.java

```diff
@@ -0,0 +1,105 @@
+
+    /**
+     * @param dataRepo   a repository storing the data of a GTFS dataset
+     * @param resultRepo a repository storing information about the validation process
+     * @param logger     a logger to log information about the validation process
+     */
+    public ValidateAgencyLangAndFeedInfoFeedLangMatch(final GtfsDataRepository dataRepo,
+                                                      final ValidationResultRepository resultRepo,
+                                                      final Logger logger) {
+        this.dataRepo = dataRepo;
+        this.resultRepo = resultRepo;
+        this.logger = logger;
+    }
+
+    /**
+     * Use case execution method: checks if for each {@code Agency} defines a value for `agency.agency_lang` that
+     * matches the value contained in `feed_info.feed_lang` if `feed_info.txt` is provided. If this requirement is not
+     * met, a {@code FeedInfoLangAgencyLangMismatchNotice} is added to the {@code ValidationResultRepo} provided
+     * in the constructor.
+     */
+    public void execute() {
+        logger.info("Validating rule 'E055 - Mismatching feed and agency language fields'");
+
+        if (dataRepo.getFeedInfoAll().size() == 0) {
+            return;
+        }
+        // .get can be used without isPresent check here since line 55 ensures the presence of at least 1 element
+        // in the map
+        final String feedInfoFeedLang =
+                dataRepo.getFeedInfoAll().values().stream().findFirst().get().getFeedLang();
+        final Set<String> agencyLangCollection = new HashSet<>();
+        dataRepo.getAgencyAll().forEach((agencyId, agency) -> agencyLangCollection.add(agency.getAgencyLang()));
+
+        dataRepo.getAgencyAll().forEach((agencyId, agency) -> {
+            if (!feedInfoFeedLang.equals(MUL)) {
+                // If feed_lang is not mul and differs from agency_lang, that's an error
+                if (!feedInfoFeedLang.equals(agency.getAgencyLang())) {
+                    resultRepo.addNotice(
+                            new FeedInfoLangAgencyLangMismatchNotice(
+                                    agencyId,
+                                    agency.getAgencyName(),
+                                    agency.getAgencyLang(),
+                                    feedInfoFeedLang));
+                    return;
+                }
+                // If there is more than one agency_lang and feed_lang isn't mul, that's an error
+                if (agencyLangCollection.size() > 1 && !feedInfoFeedLang.equals(agency.getAgencyLang())) {
+                    resultRepo.addNotice(
+                            new FeedInfoLangAgencyLangMismatchNotice(
+                                    agencyId,
+                                    agency.getAgencyName(),
+                                    agency.getAgencyLang(),
+                                    feedInfoFeedLang));
+                    return;
+                }
```

Comment:
This check should be above this loop through the agencies.txt data - it only needs to be checked once after adding all the agency_lang to `agencyLangCollection`, and should have logic like:

```java
// If there is more than one agency_lang and feed_lang isn't mul, that's an error
 if (agencyLangCollection.size() > 1 && !feedInfoFeedLang.equals(MUL))) {
                    resultRepo.addNotice(
                            new FeedInfoLangAgencyLangMismatchNotice(
                                    agencyId,
                                    agency.getAgencyName(),
                                    agency.getAgencyLang(),
                                    feedInfoFeedLang));
                    return;
}
```

There shouldn't be

Reply:
After looking at this further I think I see that you're trying to combine some of the conditionals within a single `if (!feedInfoFeedLang.equals(MUL)) {` statement, and some of those could early return on the first loop iteration, so it might be logically equivalent to moving the IFs outside the loop. However, for clarity I'd still suggest you move them above this loop, as it's a clearer indication that you only expect the logic to execute once, and not on every item in agencies.txt. I think it's also easier to match the code to the comments if the logic is the same, even if you might be repeating some of the conditionals.